webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,222 @@
+      var dbName = $window.localStorage.getItem(DB_ID_KEY);
+      if (!dbName) {
+        dbName =
+          'medic-user-' + Session.userCtx().name + '-telemetry-' + uuid();
+        $window.localStorage.setItem(DB_ID_KEY, dbName);
+      }
+      return $window.PouchDB(dbName); // avoid angular-pouch as digest isn't necessary here
+    };
+
+    var getUniqueDeviceId = function() {
+      var uniqueDeviceId = $window.localStorage.getItem(DEVICE_ID_KEY);
+      if (!uniqueDeviceId) {
+        uniqueDeviceId = uuid();
+        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
+      }
+
+      return uniqueDeviceId;
+    };
+
+    var getLastAggregatedDate = function() {
+      var date = parseInt(
+        $window.localStorage.getItem(LAST_AGGREGATED_DATE_KEY)
+      );
+      if (!date) {
+        date = Date.now();
+        $window.localStorage.setItem(LAST_AGGREGATED_DATE_KEY, date);
+      }
+      return date;
+    };
+
+    var storeIt = function(db, key, value) {
+      return db.post({
+        key: key,
+        value: value,
+        date_recorded: Date.now(),
+      });
+    };
+
+    var submitIfNeeded = function(db) {
+      var monthStart = moment().startOf('month');
+      var dbDate = moment(getLastAggregatedDate());
+      if (dbDate.isBefore(monthStart)) {
+        return aggregate(db).then(function() {
+          return reset(db);
+        });
+      }
+    };
+
+    var convertReduceToKeyValues = function(reduce) {
+      var kv = {};
+      reduce.rows.forEach(function(row) {
+        kv[row.key] = row.value;
+      });
+
+      return kv;
+    };
+
+    var generateAggregateDocId = function(metadata) {
+      return [
+        'telemetry',
+        metadata.year,
+        metadata.month,
+        metadata.user,
+        metadata.deviceId,
+      ].join('-');
+    };
+
+    var generateMetadataSection = function() {
+      var date = moment(getLastAggregatedDate());
+      return {
+        year: date.year(),
+        month: date.month(),
+        user: Session.userCtx().name,
+        deviceId: getUniqueDeviceId(),
+        // REVIEWER: is there anything else we think we should store here?
+        // Candidates would be generic metadata that is not directly device
+        // related or DB related
+      };
+    };
+
+    var generateDeviceStats = function() {
+      return {
+        userAgent: $window.navigator.userAgent,
+        language: $window.navigator.language,
+        hardwareConcurrency: $window.navigator.hardwareConcurrency,
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // REVIEWER: while we can expand this in the future are their other
+        // useful things you would like to see here?
+        //
+        // REVIEWER: do we want to ship with extra data pulled from Android via
+        // medic-android? This would allow us to know true android version, disk
+        // space, stuff like that.
```

Comment:
Yeah I think the android information is worth pulling, particularly with the performance data.